### PR TITLE
[aptos-rosetta] Add Staking balances, Set Operator and SetVoter support in Rosetta

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
This supports staking_contract and stake modules, but not vesting.

Set voter, set operator, and the balances are supported now.

### Test Plan
Testing seems to work fine currently, but we'll need to test with real data.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4450)

### CODE REVIEW

File: crates/aptos-rosetta/src/construction.rs [hunk 1]
```diff
@@ -402,11 +402,11 @@ async fn construction_parse(
                 (AccountAddress::ONE, APTOS_ACCOUNT_MODULE, CREATE_ACCOUNT_FUNCTION) => {
                     parse_create_account_operation(sender, &type_args, &args)?
                 }
-                (AccountAddress::ONE, STAKE_MODULE, SET_OPERATOR_FUNCTION) => {
```
Comment (movekevin) [inline crates/aptos-rosetta/src/construction.rs:405]: I think we should still keep the old one and make this new proxy set operator a separate call. This allows most normal use cases to use the old one, which is simpler as it doesn't need to pass the old operator.

File: crates/aptos-rosetta/src/construction.rs [hunk 2]
```diff
@@ -402,11 +402,11 @@ async fn construction_parse(
...
-                    parse_set_operator_operation(sender, &type_args, &args)?
+                (AccountAddress::ONE, STAKING_PROXY_MODULE, SET_OPERATOR_FUNCTION) => {
+                    parse_set_operator_operation(sender, &type_args, &args, None)?
                 }
-                (AccountAddress::ONE, STAKE_MODULE, SET_VOTER_FUNCTION) => {
```
Comment (movekevin) [inline crates/aptos-rosetta/src/construction.rs:408]: Same comment for the old SetVoter

File: crates/aptos-rosetta/src/construction.rs (hunk 1 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/construction.rs:405]: I think this makes the Rosetta flow a little too complicated, you need to know if you're using X you need to store the operator.  if you're using Y you don't.

File: crates/aptos-rosetta/src/construction.rs [hunk 3]
```diff
@@ -410,11 +410,11 @@ async fn construction_parse(
                 (AccountAddress::ONE, APTOS_ACCOUNT_MODULE, CREATE_ACCOUNT_FUNCTION) => {
                     parse_create_account_operation(sender, &type_args, &args)?
                 }
-                (AccountAddress::ONE, STAKE_MODULE, SET_OPERATOR_FUNCTION) => {
-                    parse_set_operator_operation(sender, &type_args, &args)?
+                (AccountAddress::ONE, STAKING_PROXY_MODULE, SET_OPERATOR_FUNCTION) => {
```
Comment (movekevin) [inline crates/aptos-rosetta/src/construction.rs:413]: If we've switched to only staking_contract, we don't really need staking proxy contract in Rosetta anymore. You can just directly call the function in staking_contract::switch_operator_with_same_commission?

**Review by movekevin — APPROVED** [review]
Subaccount was a bit hard to wrap my head around but as long as it works 👍

**Review by davidiw — COMMENTED** [review]
Looks good, should we test it with real data first before landing? I suspect this won't work right off the bat?

**Comment by gregnazario** [comment]
> Looks good, should we test it with real data first before landing? I suspect this won't work (perfectly) right off the bat?

Well, I've been mostly pushing and testing.

One of the main issues is the lack of a proper mock framework, where when I mock stuff, I don't actually get it right.

This does however have a full E2E test with test data that tests balances, staking pools, and set operator / set voter

**Review by davidiw — APPROVED** [review]

**Pull request merged**